src/compiler/scala/tools/nsc/transform/Delambdafy.scala

```diff
@@ -46,9 +48,9 @@ abstract class Delambdafy extends Transform with TypingTransformers with ast.Tre
 
     val typer = localTyper
 
-    // we need to know which methods refer to the 'this' reference so that we can determine
-    // which lambdas need access to it
-    val thisReferringMethods: Set[Symbol] = {
+    // we need to know which methods refer to the 'this' reference so that we can determine which lambdas need access to it
+    // TODO: this looks expensive, so I made it a lazy val. Can we make it more pay-as-you-go / optimize for common shapes?
```

Comment:
I think this can limit its attention to `anonfun` methods.

It would possible to accumulate the knowledge we need during the main `transform`, rather than in an additional pass. We'd need to transform template bodies in such a way that we we process definitions of `anonfun` methods before usages, which would currently amount to transforming the stats in reverse. However, if we can limit the analysis in its current form to only look at template-owned `anonfun` method bodies, it ought to be fairly low overhead.

AFAICT, part of the complexity of this analysis stems from the desire to make all the lambda impl methods static in `() => () => 42`.

https://gist.github.com/062181846c13e65490cc